File: c++/b7562.cpp
#include <iostream>
#include <deque>
#include <tuple>
#include <string.h>
using namespace std;

void BFS(deque<tuple<int, int, int>>& DQ, int field[][300], const int& I, const int& target_r, const int& target_c)
{
    pair<int, int> move[8] = {
        {2, -1}, {2, 1}, {1, -2}, {1, 2}, {-1, -2}, {-1, 2}, {-2, 1}, {-2, -1}
    };
    int now_r, now_c, step, to_r, to_c;

    while (!DQ.empty()) {
        tie(now_r, now_c, step) = DQ.front();
        DQ.pop_front();
        for (int to = 0; to < 8; to++) {
            to_r = now_r + move[to].first;
            to_c = now_c + move[to].second;
            if (0 <= to_r && to_r < I && 0 <= to_c && to_c < I && field[to_r][to_c] == 0) {
                field[to_r][to_c] = 1;
                if (to_r == target_r && to_c == target_c) {
                    cout << step + 1 << '\n';
                    DQ.clear();
                    return;
                }
                DQ.push_back(make_tuple(to_r, to_c, step + 1));
            }
        }
    }
}

int main()
{
    int T, I, start_r, start_c, target_r, target_c;
    cin >> T;
    int field[300][300];
    deque<tuple<int, int, int>> DQ;
    for (int tc = 0; tc < T; tc++) {
        memset(field, 0, sizeof(field));
        cin >> I >> start_r >> start_c >> target_r >> target_c;
        field[start_r][start_c] = 1;
        DQ.push_back(make_tuple(start_r, start_c, 0));
        if (start_r == target_r && start_c == target_c)
            cout << 0 << '\n';
        else
            BFS(DQ, field, I, target_r, target_c);
    }

    return 0;
}

File: python/b7562.py
from typing import Deque

dy = [-2, -2, -1, -1, 1, 1, 2, 2]
dx = [-1, 1, -2, 2, -2, 2, -1, 1]


def BFS():
    while queue:
        r, c, step = queue.popleft()
        for to in range(8):
            y = r + dy[to]
            x = c + dx[to]
            if 0 <= y < N and 0 <= x < N and check[y][x]:
                check[y][x] = False
                if y == target_r and x == target_c:
                    print(step + 1)
                    return
                queue.append((y, x, step + 1))


for _ in range(int(input())):
    N = int(input())
    sr, sc = map(int, input().split())
    target_r, target_c = map(int, input().split())
    if sr == target_r and sc == target_c:
        print(0)
        continue
    check = [[True] * N for _ in range(N)]
    queue = Deque()
    check[sr][sc] = False
    queue.append((sr, sc, 0))
    BFS()